### PR TITLE
feat: Style Toast component

### PR DESCRIPTION
**Pull request by lucasfp13**
## What's the purpose of this pull request?

Ported from `base.store` repo: [feat: Style Toast #496](https://github.com/vtex-sites/base.store/pull/496)

This PR aims to style the `Toast` component, animations included.

This PR is limited to animating and styling a single toast. Multiple toasts will be tackled in [another task](https://vtex-dev.atlassian.net/browse/FSSS-237?atlOrigin=eyJpIjoiMzdmNzAwNDI3ZmQyNDBkMzg2OGUxZjdhOWExMTk0NTkiLCJwIjoiaiJ9).

## How does it work?

A toast is generated by using the [pushToast](https://github.com/vtex-sites/base.store/blob/3c31e5b72c4937cc53a7f4e873ff6aad63d23a12/src/sdk/ui/index.tsx#L38) function. After displayed,  the `popToast` function will be used to hide the toast after the animation ends.

Also, toasts are being generated whenever a new notification arrives from [useCart](https://github.com/vtex-sites/base.store/blob/3c31e5b72c4937cc53a7f4e873ff6aad63d23a12/src/components/common/Toast/Toast.tsx#L20). This behavior needs to be revised since messages lack some properties such as icon and description. Maybe this is the case to use Alerts instead of Toasts.

## How to test it?

`Toast` component can be tested through Storybook.

https://user-images.githubusercontent.com/15722605/168651505-0d0767d6-3557-4916-9171-009e87ab6fd3.mov

## References

- [Jira task (FSSS-231)](https://vtex-dev.atlassian.net/browse/FSSS-231)
- [Figma prototype](https://www.figma.com/file/u9z8sTWqBm8eVGsazPLlnX/Handoff-v4-(Regionalization)?node-id=666%3A154997)

## Checklist

<em>You may erase this after checking them all ;)</em>

- [x] Added an entry in the `CHANGELOG.md` at the beginning of its due section. [The latest version should comes first](https://keepachangelog.com/en/1.0.0/#:~:text=The%20latest%20version%20comes%20first.).
- [x] Added the PR number with the PR link at the entry in the `CHANGELOG.md`. E.g., *New items in the `pull_request_template.md` ([#12](https://github.com/vtex-sites/gatsby.store/pull/12))* 

- PR description
- [x] Updated the Storybook - *if applicable*.
- [x] Added a label according to the PR goal - `Breaking change`, `Enhancement`, `Bug` or `Chore`.
- [x] Added the component, hook, or pathname in-between backticks (``) *- If applicable*. E.g., *`ComponentName` component*.

### CODE REVIEW

File: src/components/ui/Toast/toast.scss [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  // --------------------------------------------------------
+
+  // Default properties
+  --fs-toast-width                        : calc(100% - (2 * var(--fs-spacing-3)));
+
+  --fs-toast-min-height                   : var(--fs-spacing-10);
+  --fs-toast-padding                      : var(--fs-spacing-1);
+  --fs-toast-margin                       : var(--fs-spacing-3) var(--fs-spacing-3) 0 var(--fs-spacing-3);
+
+  --fs-toast-border-radius                : 8px;
```
Comment (hellofanny) [inline src/components/ui/Toast/toast.scss:15]: @renatamottam

Comment (hellofanny) [reply src/components/ui/Toast/toast.scss:15]: Let's add a new token for this value on `tokens.scss`

File: src/components/ui/Toast/toast.scss [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  --fs-toast-icon-container-width         : var(--fs-spacing-7);
+  --fs-toast-icon-container-height        : var(--fs-toast-icon-container-width);
+  --fs-toast-icon-container-bkg-color     : var(--fs-color-primary-bkg-light);
+  --fs-toast-icon-container-border-radius : var(--fs-border-radius-default);
+
+  --fs-toast-content-padding              : var(--fs-spacing-0) var(--fs-spacing-2);
+
+  --fs-toast-content-title-size           : var(--fs-text-size-body);
+  --fs-toast-content-title-weight         : var(--fs-text-weight-bold);
+  --fs-toast-content-title-line-height    : 1.25;
```
Comment (filipewl) [inline src/components/ui/Toast/toast.scss:34]: I wonder if we can use the following:

```suggestion
  --fs-toast-content-title-line-height    : var(--fs-scale);
```

CleanShot 2022-05-16 at 19 50 12

File: src/styles/global/storybook-components.scss [hunk 3]
```diff
@@ -55,4 +55,5 @@
 @import "src/components/ui/SlideOver/slide-over.scss";
 @import "src/components/ui/SROnly/sr-only.scss";
 @import "src/components/ui/Tiles/tiles.scss";
+@import "src/components/ui/Toast/toast.scss";
```
Comment (filipewl) [inline src/styles/global/storybook-components.scss:58]: If we also import it in `pages/layout.scss`, will it work for the app?

CleanShot 2022-05-16 at 19 51 53

File: src/components/ui/Toast/Toast.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import { useUI, uiActions, uiEffects, uiInitialState } from 'src/sdk/ui'
+import { UIProvider } from '@faststore/sdk'
+
+import Toast from '.'
+
+export default {
+  component: Toast,
+  title: 'Atoms/Toast',
```
Comment (filipewl) [inline src/components/ui/Toast/Toast.stories.tsx:8]: Do you know why when testing this story on mobile, clicking on the button doesn't trigger the toast?
https://gatsby-store-storybook-git-feat-fsss-231-style-e45afa-faststore.vercel.app/?path=/story/atoms-toast--default&globals=backgrounds.grid:false
CleanShot 2022-05-16 at 19 53 17

**Review by filipewl — COMMENTED** [review]
It's looking very good!

Left some comments.

File: src/components/ui/Toast/Toast.stories.tsx (hunk 4 above)
Comment (lucasfp13) [reply src/components/ui/Toast/Toast.stories.tsx:8]: maybe it just failed punctually, I tested it here through this same link and it's ok

Comment (filipewl) [reply src/components/ui/Toast/Toast.stories.tsx:8]: It was my Chrome! My bad. :~

**Review by filipewl — APPROVED** [review]

File: src/components/ui/Toast/Toast.stories.tsx (hunk 4 above)
Comment (hellofanny) [reply src/components/ui/Toast/Toast.stories.tsx:8]: I faced the same issue 👀 Not sure what happened. I'll try again.

Comment (eduardoformiga) [inline src/components/ui/Toast/Toast.stories.tsx:8]: I wonder if it should be a Molecule, what do you think?

File: src/components/ui/Toast/Toast.tsx [hunk 5]
```diff
@@ -1,35 +1,44 @@
...
+  useEffect(() => {
+    toast && setFade('in')
+  }, [toast])
+
   useEffect(() => {
-    const id = setTimeout(() => {
-      popToast()
-    }, 2e3)
+    if (fade === 'in') {
+      const id = setTimeout(() => setFade('out'), 2e3)
```
Comment (eduardoformiga) [inline src/components/ui/Toast/Toast.tsx:18]: does make sense to clear the Timeout inside the [cleanup function](https://reactjs.org/docs/hooks-effect.html#:~:text=Why%20did%20we%20return%20a%20function%20from%20our%20effect%3F) return of `useEffect`? something like:
```js
return () => {
  clearTimeout(id);
};
```

**Review by eduardoformiga — APPROVED** [review]
Good Job! 👏 

Left some minor comments.

File: src/components/ui/Toast/toast.scss [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+[data-fs-toast] {
+  // --------------------------------------------------------
+  // Design Tokens for Toast
+  // --------------------------------------------------------
+
+  // Default properties
+  --fs-toast-width                        : calc(100% - (2 * var(--fs-spacing-3)));
+
+  --fs-toast-min-height                   : var(--fs-spacing-10);
+  --fs-toast-padding                      : var(--fs-spacing-1);
```
Comment (renatamottam) [inline src/components/ui/Toast/toast.scss:12]: The padding seems off, the proposal uses a .5rem of spacing. Try to use the `--fs-spacing-1 for it` :)
|Current|Proposed|
|-|-|
|image|image|

**Review by renatamottam — APPROVED** [review]
Nice work, @lucasfp13! I suggested a few improvements below

**Comment by eduardoformiga** [comment]
Hey, another thing that I just remember now. 
Should we use `text-overflow: ellipsis` (...) for the case of long texts?
ref: https://www.w3schools.com/cssref/css3_pr_text-overflow.asp

File: src/components/ui/Toast/Toast.tsx (hunk 5 above)
Comment (lucasfp13) [reply src/components/ui/Toast/Toast.tsx:18]: humm, I think it makes sense, let me put this cleanup function and check if there's some interference on the behavior

File: src/components/ui/Toast/toast.scss (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucasfp13) [reply src/components/ui/Toast/toast.scss]: I used `toast-content...` to keep consistency with the content container data-attribute name (`data-fs-toast-content`), but I can change that

**Comment by lucasfp13** [comment]
> Hey, another thing that I just remember now. Should we use `text-overflow: ellipsis` (...) for the case of long texts? ref: https://www.w3schools.com/cssref/css3_pr_text-overflow.asp

Nice, but we need to confirm with @filipeafns and @renatamottam if we want to truncate that, because on prototype we have a multiline version like this one:

Screen Shot 2022-05-17 at 19 13 05

File: src/components/ui/Toast/toast.scss (hunk 2 above)
[the remark this answers is not in the document]
Comment (hellofanny) [reply src/components/ui/Toast/toast.scss]: I agree! I don't think we need the `content` here for those tokens.

File: src/components/ui/Toast/Toast.tsx [hunk 7]
```diff
@@ -1,35 +1,46 @@
...
 
   return (
-    <div>
-      <Button onClick={popToast}>Close</Button>
-      <div className={`h-36 text-white ${state[toast.status]}`}>
-        message: {toast.message}
+    <div
+      data-fs-toast
+      data-fs-toast-state={fade}
+      onAnimationEnd={() => fade === 'out' && popToast()}
```
Comment (hellofanny) [inline src/components/ui/Toast/Toast.tsx:34]:
```suggestion
      onAnimationEnd={() => fade === 'out' && popToast()}
      role="status"
```

Comment (hellofanny) [reply src/components/ui/Toast/Toast.tsx:34]: We should improve this component's accessibility. I'll suggest the addition of role="status"

>  content is advisory information for the user but is not important enough to justify an [alert](https://www.w3.org/TR/wai-aria-1.1/#alert), often but not necessarily presented as a status bar.
- [reference](https://www.w3.org/TR/wai-aria-1.1/#status)

I think it's not important as an alert, but the designers can confirm that.

**Comment by hellofanny** [comment]
Well done @lucasfp13 🚀 🚀 !!

I've left a suggestion, can you please check it.

I also would like two point out 2 things:

1. How long the message should stay up? I know there is many discussion related to it. Maybe, later we could leave this decision to the user,but can we increase the current one? I would suggest at [least 6s](https://sheribyrnehaber.medium.com/designing-toast-messages-for-accessibility-fb610ac364be).

2. The [text message's length ](https://github.com/vtex-sites/gatsby.store/pull/28#issuecomment-1129366826): I think that the purpose of this component is to deliver a very straightforward message and quick feedback, so maybe we can limit the max quantity of chars for this message, so we don't need/should truncate it. (If we trim it, we still need to display the full message somehow)

**Review by hellofanny — CHANGES_REQUESTED** [review]
Well done @lucasfp13 🚀 🚀 !!

I've left a suggestion, can you please check it.

I also would like two point out 2 things:

1. How long the message should stay up? I know there is many discussion related to it. Maybe, later we could leave this decision to the user,but can we increase the current one? I would suggest at [least 6s](https://sheribyrnehaber.medium.com/designing-toast-messages-for-accessibility-fb610ac364be).

2. The [text message's length ](https://github.com/vtex-sites/gatsby.store/pull/28#issuecomment-1129366826): I think that the purpose of this component is to deliver a very straightforward message and quick feedback, so maybe we can limit the max quantity of chars for this message, so we don't need/should truncate it. (If we trim it, we still need to display the full message somehow)

**Review by renatamottam — APPROVED** [review]

**Review by hellofanny — APPROVED** [review]

**Pull request merged**